File: TGPControlsDemo7/TGPControlsDemo7/ViewController.m
//    @file:    ViewController.m
//    @project: TGPControlsDemo7 (TGPControls)
//
//    @history: Created November 27, 2014 (Thanksgiving Day)
//    @author:  Xavier Schott
//              mailto://xschott@gmail.com
//              http://thegothicparty.com
//              tel://+18089383634
//
//    @license: http://opensource.org/licenses/MIT
//    Copyright (c) 2014, Xavier Schott
//
//    Permission is hereby granted, free of charge, to any person obtaining a copy
//    of this software and associated documentation files (the "Software"), to deal
//    in the Software without restriction, including without limitation the rights
//    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//    copies of the Software, and to permit persons to whom the Software is
//    furnished to do so, subject to the following conditions:
//
//    The above copyright notice and this permission notice shall be included in
//    all copies or substantial portions of the Software.
//
//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//    THE SOFTWARE.

#import "ViewController.h"
#import "TGPDiscreteSlider7.h"
#import "TGPCamelLabels7.h"

@interface ViewController ()

// iOS 7 does not support IBInspectable, so use the .7 class for both controls.
// If your project targets iOS 8+, you should use TGPDiscreteSlider & TGPCamelLabels instead

@property (weak, nonatomic) IBOutlet TGPCamelLabels7 *oneTo10Labels;
@property (weak, nonatomic) IBOutlet TGPDiscreteSlider7 *oneTo10Slider;

@property (weak, nonatomic) IBOutlet TGPCamelLabels7 *alphabetLabels;
@property (weak, nonatomic) IBOutlet TGPDiscreteSlider7 *alphabetSlider;

@property (weak, nonatomic) IBOutlet TGPCamelLabels7 * pictureLabels;
@property (weak, nonatomic) IBOutlet TGPDiscreteSlider7 * pictureSlider;

@property (weak, nonatomic) IBOutlet TGPCamelLabels7 *switch1Camel;
@property (weak, nonatomic) IBOutlet TGPCamelLabels7 *switch2Camel;

@property (weak, nonatomic) IBOutlet UILabel *controlEventsLabel;
@property (weak, nonatomic) IBOutlet TGPDiscreteSlider7 *dualColorSlider;
@property (weak, nonatomic) IBOutlet UIStepper *stepper;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.alphabetLabels.names = @[@"A",@"B",@"C",@"D",@"E",@"F", @"G",@"H",@"I",@"J",@"K",@"L",@"M",
                                  @"N",@"O",@"P",@"Q",@"R",@"S", @"T",@"U",@"V",@"W",@"X",@"Y",@"Z"];
    self.pictureLabels.names = @[@"orient", @"occident", @"zénith", @"nadir", @"septentrion", @"midi"];
    self.switch1Camel.names = @[@"OFF", @"ON"];
    self.switch2Camel.names = @[@"O", @"l"];

    // Automatically track tick spacing changes
    self.alphabetSlider.ticksListener = self.alphabetLabels;
    self.oneTo10Slider.ticksListener = self.oneTo10Labels;
    self.pictureSlider.ticksListener = self.pictureLabels;

    // UIControlEvents
    [self.dualColorSlider addTarget:self action:@selector(touchDown:event:) forControlEvents:UIControlEventTouchDown];
    [self.dualColorSlider addTarget:self action:@selector(touchDownRepeat:event:) forControlEvents:UIControlEventTouchDownRepeat];
    [self.dualColorSlider addTarget:self action:@selector(touchDragInside:event:) forControlEvents:UIControlEventTouchDragInside];
    [self.dualColorSlider addTarget:self action:@selector(touchDragOutside:event:) forControlEvents:UIControlEventTouchDragOutside];
    [self.dualColorSlider addTarget:self action:@selector(touchDragEnter:event:) forControlEvents:UIControlEventTouchDragEnter];
    [self.dualColorSlider addTarget:self action:@selector(touchDragExit:event:) forControlEvents:UIControlEventTouchDragExit];
    [self.dualColorSlider addTarget:self action:@selector(touchUpInside:event:) forControlEvents:UIControlEventTouchUpInside];
    [self.dualColorSlider addTarget:self action:@selector(touchUpOutside:event:) forControlEvents:UIControlEventTouchUpOutside];
    [self.dualColorSlider addTarget:self action:@selector(touchCancel:event:) forControlEvents:UIControlEventTouchCancel];
    [self.dualColorSlider addTarget:self action:@selector(valueChanged:event:) forControlEvents:UIControlEventValueChanged];
}

#pragma mark - UISwitch

- (IBAction)switch1ValueChanged:(UISwitch *)sender {
    [self.switch1Camel setValue:((sender.isOn) ? 1 : 0)];
}

- (IBAction)switch2TouchUpInside:(UISwitch *)sender {
    [self.switch2Camel setValue:((sender.isOn) ? 1 : 0)];
}

#pragma mark - UIControlEvents

- (void)touchDown:(UIControl *)sender event:(UIEvent *)event {
    self.controlEventsLabel.text = @"touchDown";
}
- (void)touchDownRepeat:(UIControl *)sender event:(UIEvent *)event {
    self.controlEventsLabel.text = @"touchDownRepeat";
}
- (void)touchDragInside:(UIControl *)sender event:(UIEvent *)event {
    self.controlEventsLabel.text = @"touchDragInside";
}
- (void)touchDragOutside:(UIControl *)sender event:(UIEvent *)event {
    self.controlEventsLabel.text = @"touchDragOutside";
}
- (void)touchDragEnter:(UIControl *)sender event:(UIEvent *)event {
    self.controlEventsLabel.text = @"touchDragEnter";
}
- (void)touchDragExit:(UIControl *)sender event:(UIEvent *)event {
    self.controlEventsLabel.text = @"touchDragExit";
}
- (void)touchUpInside:(UIControl *)sender event:(UIEvent *)event {
    self.controlEventsLabel.text = @"touchUpInside";
}
- (void)touchUpOutside:(UIControl *)sender event:(UIEvent *)event {
    self.controlEventsLabel.text = @"touchUpOutside";
}
- (void)touchCancel:(UIControl *)sender event:(UIEvent *)event {
    self.controlEventsLabel.text = @"touchCancel";
}
- (void)valueChanged:(TGPDiscreteSlider7 *)sender event:(UIEvent *)event {
    self.controlEventsLabel.text = @"valueChanged";
    self.stepper.value = (double) sender.value;
}

#pragma mark - UIStepper

- (IBAction)stepperValueChanged:(UIStepper *)sender {
    self.dualColorSlider.value = (CGFloat) sender.value;
}

@end

File: TGPControlsDemo/TGPControlsDemo/ViewController.swift
import UIKit
import TGPControls

class ViewController: UIViewController {
    @IBOutlet weak var oneTo10Labels: TGPCamelLabels!
    @IBOutlet weak var oneTo10Slider: TGPDiscreteSlider!

    @IBOutlet weak var alphabetLabels: TGPCamelLabels!
    @IBOutlet weak var alphabetSlider: TGPDiscreteSlider!

    @IBOutlet var pictureLabels: TGPCamelLabels!
    @IBOutlet var pictureSlider: TGPDiscreteSlider!

    @IBOutlet weak var switch1Camel: TGPCamelLabels!
    @IBOutlet weak var switch2Camel: TGPCamelLabels!

    @IBOutlet weak var controlEventsLabel: UILabel!
    @IBOutlet weak var dualColorSlider: TGPDiscreteSlider!
    @IBOutlet weak var stepper: UIStepper!

    private func localizedStrings(_ key: String) -> [String] {
        return NSLocalizedString(key, comment: "")
            .split(separator: " ")
            .map({ (substring) -> String in
                return "\(substring)"
            })
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        oneTo10Labels.names = localizedStrings("oneTo10Labels.numbers")

        alphabetLabels.names = localizedStrings("alphabetLabels.letters")
        alphabetSlider.tickCount = alphabetLabels.names.count // Number of letters in the given alphabet

        pictureLabels.names = [NSLocalizedString("pictureLabels.east", comment: ""),
                               NSLocalizedString("pictureLabels.west", comment: ""),
                               NSLocalizedString("pictureLabels.up", comment: ""),
                               NSLocalizedString("pictureLabels.down", comment: ""),
                               NSLocalizedString("pictureLabels.north", comment: ""),
                               NSLocalizedString("pictureLabels.south", comment: "")]

        switch1Camel.names = [NSLocalizedString("switch1Camel.off", comment: ""),
                              NSLocalizedString("switch1Camel.on", comment: "")]

        switch2Camel.names = [NSLocalizedString("switch2Camel.off", comment: ""),
                              NSLocalizedString("switch2Camel.on", comment: "")]


        // Automatically track tick spacing changes and UIControlEventValueChanged
        alphabetSlider.ticksListener = alphabetLabels
        oneTo10Slider.ticksListener = oneTo10Labels
        pictureSlider.ticksListener = pictureLabels

        // UIControlEvents
        dualColorSlider.addTarget(self, action: #selector(ViewController.touchDown(_:event:)), for: .touchDown)
        dualColorSlider.addTarget(self, action: #selector(ViewController.touchDownRepeat(_:event:)), for: .touchDownRepeat)
        dualColorSlider.addTarget(self, action: #selector(ViewController.touchDragInside(_:event:)), for: .touchDragInside)
        dualColorSlider.addTarget(self, action: #selector(ViewController.touchDragOutside(_:event:)), for: .touchDragOutside)
        dualColorSlider.addTarget(self, action: #selector(ViewController.touchDragEnter(_:event:)), for: .touchDragEnter)
        dualColorSlider.addTarget(self, action: #selector(ViewController.touchDragExit(_:event:)), for: .touchDragExit)
        dualColorSlider.addTarget(self, action: #selector(ViewController.touchUpInside(_:event:)), for: .touchUpInside)
        dualColorSlider.addTarget(self, action: #selector(ViewController.touchUpOutside(_:event:)), for: .touchUpOutside)
        dualColorSlider.addTarget(self, action: #selector(ViewController.touchCancel(_:event:)), for: .touchCancel)
        dualColorSlider.addTarget(self, action: #selector(ViewController.valueChanged(_:event:)), for: .valueChanged)
    }

    // MARK: - UISwitch

    @IBAction func switch1ValueChanged(_ sender: UISwitch) {
        switch1Camel.value = (sender.isOn) ? 1 : 0
    }

    @IBAction func switch2TouchUpInside(_ sender: UISwitch) {
        switch2Camel.value = (sender.isOn) ? 1 : 0
    }

    // MARK: - UIControlEvents

    @objc func touchDown(_ sender: UIControl, event:UIEvent) {
        controlEventsLabel.text = "touchDown"
    }
    @objc func touchDownRepeat(_ sender: UIControl, event:UIEvent) {
        controlEventsLabel.text = "touchDownRepeat"
    }
    @objc func touchDragInside(_ sender: UIControl, event:UIEvent) {
        controlEventsLabel.text = "touchDragInside"
    }
    @objc func touchDragOutside(_ sender: UIControl, event:UIEvent) {
        controlEventsLabel.text = "touchDragOutside"
    }
    @objc func touchDragEnter(_ sender: UIControl, event:UIEvent) {
        controlEventsLabel.text = "touchDragEnter"
    }
    @objc func touchDragExit(_ sender: UIControl, event:UIEvent) {
        controlEventsLabel.text = "touchDragExit"
    }
    @objc func touchUpInside(_ sender: UIControl, event:UIEvent) {
        controlEventsLabel.text = "touchUpInside"
    }
    @objc func touchUpOutside(_ sender: UIControl, event:UIEvent) {
        controlEventsLabel.text = "touchUpOutside"
    }
    @objc func touchCancel(_ sender: UIControl, event:UIEvent) {
        controlEventsLabel.text = "touchCancel"
    }
    @objc func valueChanged(_ sender: TGPDiscreteSlider, event:UIEvent) {
        controlEventsLabel.text = "valueChanged"
        stepper.value = Double(sender.value)
    }

    // MARK: - UIStepper

    @IBAction func stepperValueChanged(_ sender: UIStepper) {
        dualColorSlider.value = CGFloat(sender.value)
    }
}

